frame/pools/src/lib.rs

```diff
@@ -0,0 +1,863 @@
+//! Delegation pools for nominating in `pallet-staking`.
+//!
+//! The delegation pool abstraction is concretely composed of:
+//!
+//! * primary pool: This pool represents the actively staked funds ...
+//! * rewards pool: The rewards earned by actively staked funds. Delegator can withdraw rewards once
+//! * sub pools: This a group of pools where we have a set of pools organized by era
+//!   (`SubPoolsContainer.with_era`) and one pool that is not associated with an era
+//!   (`SubsPoolsContainer.no_era`). Once a `with_era` pool is older then `current_era -
+//!   MaxUnbonding`, its points and balance get merged into the `no_era` pool.
+//!
+//! # Joining
+//!
+//! # Claiming rewards
+//!
+//! # Unbonding and withdrawing
+//!
+//! # Slashing
+//!
+//! # Pool creation
+//!
+//! # Negatives
+//! - no voting
+//! - ..
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::{
+	ensure,
+	pallet_prelude::*,
+	storage::bounded_btree_map::BoundedBTreeMap,
+	traits::{Currency, ExistenceRequirement, Get},
+	DefaultNoBound,
+};
+use scale_info::TypeInfo;
+use sp_arithmetic::{FixedPointNumber, FixedU128};
+use sp_runtime::traits::{Convert, One, Saturating, StaticLookup, TrailingZeroInput, Zero};
+use sp_staking::{EraIndex, PoolsInterface, StakingInterface};
+use sp_std::collections::btree_map::BTreeMap;
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+pub use pallet::*;
+pub(crate) const LOG_TARGET: &'static str = "runtime::pools";
+
+// Syntactic sugar for logging.
+#[macro_export]
+macro_rules! log {
+	($level:tt, $patter:expr $(, $values:expr)* $(,)?) => {
+		log::$level!(
+			target: LOG_TARGET,
+			concat!("[{:?}] 💰", $patter), <frame_system::Pallet<T>>::block_number() $(, $values)*
+		)
+	};
+}
+
+type PoolId = u32;
+type BalanceOf<T> =
+	<<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+type PointsOf<T> = BalanceOf<T>;
+type SubPoolsWithEra<T> = BoundedBTreeMap<EraIndex, UnbondPool<T>, <T as Config>::MaxUnbonding>;
+
+const POINTS_TO_BALANCE_INIT_RATIO: u32 = 1;
+
+#[derive(Encode, Decode, MaxEncodedLen, TypeInfo)]
+#[cfg_attr(feature = "std", derive(frame_support::DebugNoBound, Clone, PartialEq))]
```

Comment:
```suggestion
#[derive(Encode, Decode, MaxEncodedLen, TypeInfo, frame_support::RuntimeDebugNoBound)]
#[cfg_attr(feature = "std", derive(Clone, PartialEq))]
```

Debug is quite fine to have in the runtime as well.